lsm.go

```diff
@@ -0,0 +1,352 @@
+// The Compact function is called when the level reaches it's max size. NOTE: that this is not yet implemtened and the database will rotate the full block as normal.
+type LevelConfig struct {
+	Level   SentinelType
+	MaxSize int64
+	Compact func([]*parts.Part) (*parts.Part, int64, int64, error)
+}
+
+type LSMOption func(*LSM)
+
+func LSMWithRegistry(reg prometheus.Registerer) LSMOption {
+	return func(l *LSM) {
+		l.metrics.reg = reg
+	}
+}
+
+func LSMWithLogger(logger log.Logger) LSMOption {
+	return func(l *LSM) {
+		l.logger = logger
+	}
+}
+
+// NewLSM returns an LSM-like index of len(levels) levels.
+func NewLSM(prefix string, levels []*LevelConfig, options ...LSMOption) (*LSM, error) {
+	if err := validateLevels(levels); err != nil {
+		return nil, err
+	}
+
+	lsm := &LSM{
+		prefix:     prefix,
+		levels:     NewList(L0),
+		sizes:      make([]*atomic.Int64, len(levels)),
+		configs:    levels,
+		compacting: &atomic.Bool{},
+		logger:     log.NewNopLogger(),
+		metrics: &LSMMetrics{
+			reg: prometheus.NewRegistry(),
+		},
+	}
+
+	for _, opt := range options {
+		opt(lsm)
+	}
+
+	for i := range lsm.sizes {
+		lsm.sizes[i] = &atomic.Int64{}
+	}
+
+	for i := len(levels) - 1; i > 0; i-- {
+		lsm.levels.Sentinel(levels[i].Level)
+	}
+
+	lsm.metrics.Compactions = promauto.With(lsm.metrics.reg).NewCounterVec(prometheus.CounterOpts{
+		Name: "frostdb_lsm_compactions_total",
+		Help: "The total number of compactions that have occurred.",
+	}, []string{"level"})
+
+	lsm.metrics.LevelSize = promauto.With(lsm.metrics.reg).NewGaugeVec(prometheus.GaugeOpts{
+		Name: "frostdb_lsm_level_size_bytes",
+		Help: "The size of the level in bytes.",
+	}, []string{"level"})
+
+	lsm.metrics.CompactionDuration = promauto.With(lsm.metrics.reg).NewHistogram(prometheus.HistogramOpts{
+		Name:    "frostdb_lsm_compaction_total_duration_seconds",
+		Help:    "Total compaction duration",
+		Buckets: prometheus.ExponentialBuckets(0.5, 2, 25),
+	})
+
+	return lsm, nil
+}
+
+func validateLevels(levels []*LevelConfig) error {
+	for i, l := range levels {
+		if int(l.Level) != i {
+			return fmt.Errorf("level %d is not in order", l.Level)
+		}
+
+		switch i {
+		case len(levels) - 1:
+			if l.Compact != nil {
+				return fmt.Errorf("level %d is the last level and should not have a compact function", l.Level)
+			}
+		default:
+			if l.Compact == nil {
+				return fmt.Errorf("level %d is not the last level and should have a compact function", l.Level)
+			}
+		}
+	}
+
+	return nil
+}
+
+func (l *LSM) MaxLevel() SentinelType {
+	return SentinelType(len(l.configs) - 1)
+}
+
+func (l *LSM) Add(tx uint64, record arrow.Record) {
+	record.Retain()
+	size := util.TotalRecordSize(record)
+	l.levels.Prepend(parts.NewArrowPart(tx, record, int(size), nil))
+	l0 := l.sizes[L0].Add(int64(size))
+	l.metrics.LevelSize.WithLabelValues(L0.String()).Set(float64(l0))
+	if l0 >= l.configs[L0].MaxSize {
+		go l.compact()
```

Comment:
Not sure if you're already working on this, but this reintroduces the problem of spawning a lot of goroutines for compaction which is expensive. Are you planning to reintroduce sweep-style compaction?

At the very least, I would check the atomic before spawning a goroutine.